framework/components/s3provider/minio.go

```diff
@@ -0,0 +1,241 @@
+	tc "github.com/testcontainers/testcontainers-go"
+	tcwait "github.com/testcontainers/testcontainers-go/wait"
+	"math/rand"
+	"os"
+	"strconv"
+)
+
+const (
+	DefaultImage       = "minio/minio"
+	DefaultName        = "minio"
+	DefaultBucket      = "test-bucket"
+	DefaultRegion      = "us-east-1"
+	DefaultPort        = 9000
+	DefaultConsolePort = 9001
+)
+
+type Minio struct {
+	host        string
+	port        int
+	consolePort int
+	accessKey   string
+	secretKey   string
+	bucket      string
+	region      string
+	keep        bool
+}
+
+func (m Minio) GetSecretKey() string {
+	return m.secretKey
+}
+
+func (m Minio) GetAccessKey() string {
+	return m.accessKey
+}
+
+func (m Minio) GetBucket() string {
+	return m.bucket
+}
+
+func (m Minio) GetConsoleURL() string {
+	return fmt.Sprintf("http://%s:%d", m.host, m.consolePort)
+}
+
+func (m Minio) GetEndpoint() string {
+	return fmt.Sprintf("%s:%d", m.host, m.port)
+}
+
+func (m Minio) GetRegion() string {
+	return m.region
+}
+
+type Option func(*Minio)
+
+type MinioFactory struct{}
+
+func NewMinioFactory() ProviderFactory {
+	return MinioFactory{}
+}
+
+func (mf MinioFactory) New(options ...Option) (Provider, error) {
+	m := &Minio{
+		port:        DefaultPort,
+		consolePort: DefaultConsolePort,
+		accessKey:   randomStr(20),
+		secretKey:   randomStr(40),
+		bucket:      DefaultBucket,
+		region:      DefaultRegion,
+		keep:        false,
+	}
+
+	for _, opt := range options {
+		opt(m)
+	}
+
+	var tcRyukDisabled string
+	if m.keep {
+		// store original env var to value
+		tcRyukDisabled = os.Getenv("TESTCONTAINERS_RYUK_DISABLED")
+		err := os.Setenv("TESTCONTAINERS_RYUK_DISABLED", "true")
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	ctx := context.Background()
+	containerName := framework.DefaultTCName(DefaultName)
+	bindPort := fmt.Sprintf("%d/tcp", m.port)
+	bindConsolePort := fmt.Sprintf("%d/tcp", m.consolePort)
+	networks := []string{"compose_default"}
+	networkAliases := map[string][]string{
+		"compose_default": {DefaultName},
+	}
+
+	if len(framework.DefaultNetworkName) == 0 {
+		// attach default ctf network if initiated
+		networks = append(networks, framework.DefaultNetworkName)
+		networkAliases[framework.DefaultNetworkName] = []string{
+			containerName,
+			DefaultName,
+		}
+	}
+
+	req := tc.ContainerRequest{
+		Name:           containerName,
+		Image:          DefaultImage,
+		Labels:         framework.DefaultTCLabels(),
+		Networks:       networks,
+		NetworkAliases: networkAliases,
+		ExposedPorts: []string{
+			bindPort,
+			bindConsolePort,
+		},
+		Env: map[string]string{
+			"MINIO_ROOT_USER":     m.accessKey,
+			"MINIO_ROOT_PASSWORD": m.secretKey,
+			"MINIO_BUCKET":        DefaultBucket,
+		},
+		Entrypoint: []string{
+			"minio",
+			"server",
+			"/data",
+			"--address",
+			fmt.Sprintf(":%d", m.port),
+			"--console-address",
+			fmt.Sprintf(":%d", m.consolePort),
+		},
+		HostConfigModifier: func(h *container.HostConfig) {
+			framework.NoDNS(true, h)
```

Comment:
This method is only needed for components that may connect to real staging infrastructure using DNS, we can remove it in Minio case, I think